### PR TITLE
add yaml support with pocket constraints and bond constraints

### PR DESCRIPTION
**Pull request by hgbrian**
Hey Nick,
as we discussed, I had a stab at using yaml instead of fasta.

The main benefit is the ability to supply constraints as well as sequences.

The code also allows for multiple target sequences, which I think is a common use-case and doesn't add much complexity to the code.

There are a few warts:

- the old version did not allow a custom msa, so i did not include that either
- the yaml requires a special formatter to match Boltz
- it takes a `binder_sequence` (or `binder_len`, optionally), but does not really have other code necessary to take advantage of that (I use a partially masked binder_sequence quite a lot)
- it flexibly takes one target sequence, or multiple, but the code is a bit long here. I could see forcing just taking lists here instead.
- I tested it but not super-extensively

Here is a minimalish example of using it (adapted from `example_notebook.py`)

```python
from pathlib import Path
import jax
import equinox as eqx
import numpy as np
import optax
import boltz
from boltz_binder_design import design_bregman_optax
from boltz_binder_design.losses.boltz import (
    make_binder_features,
    get_input_yaml,
    load_features_and_structure_writer,
    set_binder_sequence,
    WithinBinderContact,
    BinderTargetContact,
    StructurePrediction,
    load_boltz_model,
)

target_sequence = "SFPASVQLHT"
model = load_boltz_model()
j_model = eqx.filter_jit(lambda model, *args, **kwargs: model(*args, **kwargs))


def predict(sequence, features):
    o = j_model(
        model,
        set_binder_sequence(sequence, features),
        key=jax.random.key(5),
        sample_structure=True,
        confidence_prediction=True,
        deterministic=True,
    )
    print("plddt", o["plddt"][: sequence.shape[0]].mean())
    print("ipae", o["complex_ipae"].item())
    return o


scaffold_sequence = "SVIEKLRKLESVIEKLRKLESASPER"
binder_length = len(scaffold_sequence)
boltz_features, boltz_writer = make_binder_features(
    binder_length,
    target_sequence,
)

loss = StructurePrediction(
    model=model,
    name="target",
    loss=2 * BinderTargetContact() + WithinBinderContact(),
    features=boltz_features,
    recycling_steps=0,
    deterministic=False,
)

_, logits = design_bregman_optax(
    loss_function=loss,
    n_steps=1,
    x=np.random.randn(binder_length, 20) * 0.1,
    optim=optax.chain(
        optax.clip_by_global_norm(1.0),
        optax.sgd(0.5 * np.sqrt(binder_length), momentum=0.5),
    ),
)


# Let's repredict our designed sequence with the correct sidechains, hopefully Boltz still likes it
def repredict(logits, target_sequence, pocket_constraints, bond_constraints):
    binder_seq = "".join(
        boltz.data.const.prot_token_to_letter[boltz.data.const.tokens[i]]
        for i in logits.argmax(-1) + 2
    )

    out_dir = Path(f"/tmp/proteins/{binder_seq[:10]}_{target_sequence[:10]}")
    out_dir.mkdir(exist_ok=True, parents=True)

    yaml_path = out_dir / "protein.yaml"
    yaml_path.write_text(
        get_input_yaml(
            binder_sequence=binder_seq,
            binder_use_msa=True,
            targets_sequence=target_sequence,
            pocket_constraints=pocket_constraints,
            bond_constraints=bond_constraints,
        )
    )

    return load_features_and_structure_writer(yaml_path, out_dir)


test_pocket_constraints = [("B", 5), ("B", 6)]
test_bond_constraints = [{"atom1": ("A", 1, "CA"), "atom2": ("B", 5, "N")}]

f_r, _w = repredict(
    logits,
    target_sequence=target_sequence,
    pocket_constraints=test_pocket_constraints,
    bond_constraints=test_bond_constraints,
)

repredicted_output = predict(f_r["res_type"][0][:, 2:22], f_r)

print("repredicted_output", repredicted_output)
```

### CODE REVIEW

**Comment by nboyd** [comment]
This is super nice, thank you Brian! 

Will merge immediately but I might change a few things slightly over time!

In terms of masked `binder_sequence,` do you mean fixing certain positions to e.g. follow the wild type? If so I've been doing the following, which I think should work well with your `binder_sequence` stuff (I think ideally you'd have a sequence with `X`'s in the positions you're optimizing and WT AAs elsewhere).


```
class FixedPositionsPenalty(LossTerm):
    """Squared error for fixed positions."""

    position_mask: Bool[Array, "N"]
    target: Float[Array, "N 20"]

    def __call__(self, seq: Float[Array, "N 20"], *, key):
        r = (((seq - self.target) ** 2).sum(-1) * self.position_mask).sum()

        return r, {"fixed_position_penalty": r}

    @staticmethod
    def from_residues(
        sequence_length: int, positions_and_AAs: list[tuple[int, str]]
    ):
        position_mask = np.zeros(sequence_length, dtype=bool)
        target = np.zeros((sequence_length, len(TOKENS)))
        for idx, AA in positions_and_AAs:
            position_mask[idx] = True
            target[idx, TOKENS.index(AA)] = 1.0

        return FixedPositionsPenalty(
            jnp.array(position_mask), jnp.array(target)
        )
```


```
class ComposeFixPositions(LossTerm):
    """Precomposes loss functional with function that fixes positions.

    WARNING: Logits returned by optimizer in fixed positions will be random! Be sure to call `get_sequence` *after* optimization, e.g. `loss.get_sequence(jax.nn.softmax(logits))`"""

    positions_and_AAs: list[tuple[int, str]]
    loss: LossTerm

    def __call__(self, seq: Float[Array, "N 20"], *, key):
        return self.loss(self.get_sequence(seq), key=key)

    def get_sequence(self, seq: Float[Array, "N 20"]):
        for idx, AA in self.positions_and_AAs:
            seq = seq.at[idx].set(0.0).at[idx, TOKENS.index(AA)].set(1)
        return seq
```

The first option is nice because it's a soft penalty but might affect the optimization, the second will definitely fix those positions without affecting optimization but there's a minor footgun in that the returned logits will be essentially random in the fixed positions.

**Comment by hgbrian** [comment]
Hi Nick,
of course feel free to fix it up to match your style! Hopefully it's not buggy...

Re: binder_sequence, yes I have been providing wild-type amino acids and Xs where I want to allow exploration. I think this is probably a very common use-case in binder design so it might be worthwhile including the code you have above in your main code.

Thanks for the fixed sequence code!
I have been using a soft penalty that in practice does not seem to allow changes.
Just FYI below, but I will switch over to your method since I think it's cleaner to just another loss...

```python
    def initialize_sequence_logits(
        binder_seq: str,
        binder_seq_mask: np.ndarray[bool] | None = None,
        logit_val: float = 100.0,
    ) -> np.ndarray:
        """Initializes logits based on the provided binder sequence
        Note - 2 skips tokens <pad> and -.
        Set uniform logits for mask=False areas.
        """
        logits = np.zeros((len(binder_seq), 20))
        for i, aa in enumerate(binder_seq):
            if aa != "X":
                aa_token = boltz.data.const.prot_letter_to_token[aa]
                aa_index = boltz.data.const.tokens.index(aa_token) - 2
                logits[i, aa_index] = logit_val
            else:
                logits[i, :] = 0.0

        # Set random low values for unmasked (Xs)
        if binder_seq_mask is not None:
            logits[~binder_seq_mask, :] = (
                np.random.randn(np.sum(~binder_seq_mask), 20) * 0.1
            )

        return logits
```

**Comment by nboyd** [comment]
Oh wow this is a super good idea for initialization, I might do this as well as adding a penalty! 

p.s. FYI I added a standard token order [here](https://github.com/escalante-bio/boltz-binder-design/blob/f063575bdf0021b6e5a59a7b037e86afc92b6708/src/boltz_binder_design/common.py#L5), the manual +/- 2 kept tripping me up

**Comment by hgbrian** [comment]
Thanks! I will start using TOKENS instead.
